sklearn_numba_dpex/kmeans/drivers.py

```diff
@@ -0,0 +1,393 @@
+            global_mem_cache_size=self.global_mem_cache_size,
+            centroids_window_width_multiplier=self.centroids_window_width_multiplier,
+            centroids_window_height=self.centroids_window_height,
+            centroids_private_copies_max_cache_occupancy=self.centroids_private_copies_max_cache_occupancy,
+            work_group_size=work_group_size,
+            dtype=dtype,
+        )
+
+        reset_centroid_counts_private_copies = make_initialize_to_zeros_2dim_kernel(
+            n_samples=n_clusters,
+            n_features=nb_centroids_private_copies,
+            work_group_size=work_group_size,
+            dtype=dtype,
+        )
+
+        reset_centroids_private_copies = make_initialize_to_zeros_3dim_kernel(
+            dim0=nb_centroids_private_copies,
+            dim1=n_features,
+            dim2=n_clusters,
+            work_group_size=work_group_size,
+            dtype=dtype,
+        )
+
+        reduce_centroid_counts_private_copies = make_sum_reduction_2dim_kernel(
+            n_samples=n_clusters,
+            n_features=nb_centroids_private_copies,
+            work_group_size=work_group_size,
+            dtype=dtype,
+        )
+
+        reduce_centroids_private_copies = make_sum_reduction_3dim_kernel(
+            dim0=nb_centroids_private_copies,
+            dim1=n_features,
+            dim2=n_clusters,
+            work_group_size=work_group_size,
+            dtype=dtype,
+        )
+
+        fixed_window_assignment_kernel = make_assignment_fixed_window_kernel(
+            n_samples,
+            n_features,
+            n_clusters,
+            preferred_work_group_size_multiple=self.preferred_work_group_size_multiple,
+            centroids_window_width_multiplier=self.centroids_window_width_multiplier,
+            centroids_window_height=self.centroids_window_height,
+            work_group_size=work_group_size,
+            dtype=dtype,
+        )
+
+        # TODO: let the user pass directly dpctl.tensor arrays to avoid copies.
+        if self.X_layout == "C":
+            # TODO: support the C layout and benchmark it and default to it if
+            # performances are better
+            raise ValueError("C layout is currently not supported.")
+            X_t = dpctl.tensor.from_numpy(X, device=self.device).T
+            assert (
+                X_t.strides[0] == 1
+            )  # Fortran memory layout, equivalent to C layout on transposed
+        elif self.X_layout == "F":
+            X_t = dpctl.tensor.from_numpy(X.T, device=self.device)
+            assert (
+                X_t.strides[1] == 1
+            )  # C memory layout, equivalent to Fortran layout on transposed
+        else:
+            raise ValueError(
+                f"Expected X_layout to be equal to 'C' or 'F', but got {self.X_layout} ."
+            )
+
+        centroids_t = dpctl.tensor.from_numpy(
+            centers_init.astype(np.float32).T, device=self.device
```

Comment:
This should be `X.dtype` here. Otherwise, the computation always uses float32 operations.